File: regression/단순회귀/regression-simplehg.R

# 1. 데이터를 로드합니다.
hg <- read.csv("simple_hg.csv", header = T)
head(hg)


# 2. 데이터를 산포도 그래프로 시각화합니다.
plot(hg$input ~ hg$cost, data = hg, pch = 21, col = 'blue', bg = 'blue')



# 3. 회귀분석을 해서 회귀계수인 기울기와 절편을 구합니다.

model <- lm(hg$input ~ hg$cost, data = hg)
model

#
# Coefficients:
#   (Intercept)      hg$cost  
# 62.929        2.186 

# 4. 2번에서 시각화한 산포도 그래프에 회귀직선을 겹처서 그립니다.

abline(model, col = 'red')


# 5. 그래프 제목을 회귀직선의 방정식으로 출력되게 합니다.

title( paste( '매출 =',model$coefficients[2], 'x  광고비 + ', model$coefficients[1]  ) )




# 6. 오차 그리는 코드

y_hat <- predict( model, cost = hg$cost )       #  input 매출액 예측값 출력 ( 직선 그래프의 값 )
y_hat                                                                    # 15개의 예측값 출력

join <- function(i){                      # join 이라는 이름의 함수 생성
  lines( c( hg$cost[i], hg$cost[i]), c( hg$input[i], y_hat[i] ), col = 'green' )     # 녹색라인그래프
}

sapply(1:19, join)                # 1:19들을 join 함수에 적용하는 것 , 판다스의 apply나 map같은 느낌?




# ※ 오차와 잔차와의 차이 ?
  
# 1. 오차 : 모집단에서 실제값이 회귀선과 비교했을때의 차이 ( 실제값과 예측값과의 차이 )
# 2. 잔차 : 표본에서 실제값과 회귀선과 비교했을때의 차이 ( 실제값과 예측값과의 차이 )

File: regression/단순회귀/regression-simplehg.py
# -*- coding: utf-8 -*-
"""
Created on Thu Dec 16 16:49:45 2021

@author: parkk
"""

hg = pd.read_csv("c:\\data\\simple_hg.csv")

x = hg[['cost']]         # 독립변수
y = hg[['input']]        # 종속변수

from sklearn.linear_model import LinearRegression

model = LinearRegression()

model.fit(x, y)

print('기울기 :', model.coef_ )
print('절편 :',model.intercept_ )

#
# 기울기 : [[2.18648985]]
# 절편 : [62.92913386]
# 매출액 = 2.1864*광고비 + 62.92913

result = model.predict([[26]])    
print(result)                                  # [[119.77786987]] , 11.9억

y_hat = model.predict(x)               # 회귀직선에 넣고 예측한값

import matplotlib.pyplot as plt       # 그래프 그리기위한 모듈
import seaborn as sns

plt.figure( figsize = (10,5) )               # 그래프 사이즈 가로 10, 세로 5
ax1 = sns.distplot( y, hist = False, label = 'y', color = 'red' )          # 실제 값을 라인그래프로 시각화
ax2 = sns.distplot( y_hat, hist= False, label = 'y_hat', ax = ax1, color ='blue' )  # 예측값을 라인그래프로 시각화
plt.show()
plt.close()




r_square = model.score( x,y )
print(r_square)                           # 0.7884035286357817

sns.regplot(x,y)







